### PR TITLE
Subprocess queueing and pooling management

### PR DESCRIPTION
**Pull request by matthewrmshin**
Replace the use of multiprocessing pool with in-house logic to manage
queueing and launching of external commands, and of running
sub-processes.

Avoid creating unused child processes. Avoid having to start up pool
early.

Adjustable main loop sleep time.

(This will facilitate #715 and #2315, but likely to conflict badly with #2423.)

### CODE REVIEW

**Comment by matthewrmshin** [comment]
(Initial performance test using the complex suite indicates a higher CPU usage, but improvement to both memory and elapsed time.)

File: lib/cylc/mp_pool.py [hunk 1]
```diff
@@ -142,136 +111,176 @@ def __str__(self):
...
+        # Wait for child processes
+        self.process()
 
+    @classmethod
+    def _run_command_init(cls, ctx, callback=None, callback_args=None):
+        """Execute a shell command and capture its output and exit status."""
+        try:
+            if ctx.cmd_kwargs.get('stdin_file_paths'):
+                if len(ctx.cmd_kwargs['stdin_file_paths']) > 1:
+                    stdin_file = TemporaryFile()
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/mp_pool.py]: :disappointed: Sadly, `StringIO` does not appear to work with `subprocess` in my environment. Get a traceback with `AttributeError: StringIO instance has no attribute 'fileno'`.

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply lib/cylc/mp_pool.py]: Can we just provide the files list a string using `Popen.communicate`:

```python
>>> Popen(["sed", 's/cl/lc/'], stdout=PIPE, stdin=PIPE).communicate('cycl')[0]
cylc
```

[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/mp_pool.py]: Not good because `communicate` can block.

[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/mp_pool.py]: From the doc for `Popen.communicate`:

> ... Wait for process to terminate. ...

> The data read is buffered in memory, so do not use this method if the data size is large or unlimited.

We cannot wait for the process to terminate here. If we are really keen, however, we can write directly to `proc.stdin` (while reading from `proc.stdout` and `proc.stdout`) and use `select.select` to determine if we can write to it or not, and handle all the buffering, etc. However, I think it is probably too much hassle for what we are doing here for now.

(For reference, we have got a Perl version of this sort of logic in [FCM::Util::Shell](https://github.com/metomi/fcm/blob/master/lib/FCM/Util/Shell.pm), so if we have a strong requirement here, it should be possible to implement something similar in Python.)

**Comment by matthewrmshin** [comment]
(Re-based and resolved conflicts.)

File: tests/events/07-task-iso.t [hunk 2]
```diff
@@ -1,32 +0,0 @@
...
-#
-# You should have received a copy of the GNU General Public License
-# along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#-------------------------------------------------------------------------------
-# Validate and run the task events suite (iso mode).
-. $(dirname $0)/test_header
-#-------------------------------------------------------------------------------
-set_test_number 2
-#-------------------------------------------------------------------------------
-install_suite $TEST_NAME_BASE task-iso
```
Comment (oliver-sanders) [inline tests/events/07-task-iso.t]: The `task-iso` suite is still present.

Comment (matthewrmshin) [reply tests/events/07-task-iso.t]: Now removed.

**Comment by hjoliver** [comment]
> likely to conflict badly with #2423.

It sure did.

**Pull request merged**